### PR TITLE
[Refactoring] Enhance GotoDeclaration to support multiple open solutions...

### PR DESCRIPTION
**Pull request by sgmunn**
....

If you have more than one open solution, using GotoDeclaration for a type in the other, non-active solution, results in that type being displayed in the Assembly Browser. This change changes that behaviour to search for the type in any of the other open solutions and open the source file for the type instead.

### CODE REVIEW

**Comment by slluis** [comment]
@sgmunn @mkrueger is this PR still valid?

**Comment by mkrueger** [comment]
The question is in general if we should try to support what roslyn doesn't support. For goto declaration I would take the vs.net commands sooner or later and that functionality would break then.

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/GotoDeclarationHandler.cs [hunk 1]
```diff
@@ -48,12 +54,88 @@ protected override void Run (object data)
...
+		}
+
+		/// <summary>
+		/// Checks if the INamedElement is actually defined in another open solution and returns a INamedElement from that solution instead.
+		/// Provides support for going to the source code definition of an element instead of the Assembly Browser.
+		/// Returns the original entity if no suitable open solution was found
+		/// </summary>
+		static INamedElement CheckIfDefinedInOtherOpenSolution (INamedElement entity)
+		{
+			var ex = entity as IEntity;
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/GotoDeclarationHandler.cs]: ex ? We use that identifier for exceptions. It's not a very good choice IMO.

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/GotoDeclarationHandler.cs [hunk 2]
```diff
@@ -48,12 +54,88 @@ protected override void Run (object data)
...
+
+			if (ex != null && ex.Region.IsEmpty) {
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/GotoDeclarationHandler.cs]: Can you return here and in the next if instead - try to not get too deep if it's not needed.

File: main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/GotoDeclarationHandler.cs [hunk 3]
```diff
@@ -48,12 +54,88 @@ protected override void Run (object data)
...
+						var monitor = IdeApp.Workbench.ProgressMonitors.GetSearchProgressMonitor (true, true);
+						using (monitor) {
+							monitor.BeginTask (GettextCatalog.GetString ("Building type graph in solution ..."), 1); 
+
+							var assemblies = GetAllAssemblies (solution);
+							var tg = new TypeGraph (assemblies);
+							var node = tg.GetNode (ex.DeclaringTypeDefinition); 
+							if (node != null) {
+								// look for the member that we're after
+								var foundMember = node.TypeDefinition.Members.FirstOrDefault (x => x.FullName == entity.FullName);
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.Refactoring/MonoDevelop.Refactoring/GotoDeclarationHandler.cs]: Avoid LINQ - this is very costly.

**Comment by sgmunn** [comment]
It’s not clear to me if we still want this. I’m happy to update it if we do.

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.6.0.128 (master)

Pull Request #8937 merged by: Vsevolod Kukol
Author: mono
Commit: 86136a7ce63253563bc2ae7bbb2881ae53d92616 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.4.0.2272 (release-8.4)

Pull Request #9165 merged by: Vsevolod Kukol
Author: mono
Commit: 1f9933742db5428e56dea24e6aefb5a5c128d66b (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.6.0.128 (master)

Pull Request #8937 merged by: Vsevolod Kukol
Author: mono
Commit: 86136a7ce63253563bc2ae7bbb2881ae53d92616 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.4.0.2272 (release-8.4)

Pull Request #9165 merged by: Vsevolod Kukol
Author: mono
Commit: 1f9933742db5428e56dea24e6aefb5a5c128d66b (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.5.0.1066 (release-8.5)

Pull Request #8937 merged by: Vsevolod Kukol
Author: mono
Commit: 86136a7ce63253563bc2ae7bbb2881ae53d92616 (mono/monodevelop)

**Pull request closed without merging**